### PR TITLE
Allow specifying AdaGrad initial accumulator value

### PR DESCRIPTION
**Pull request by leezu**
## Description ##
The mxnet AdaGrad implementation currently initializes the state to 0. As the
AdaGrad update normalizes the gradient by the history value, a non-zero history
initialization is useful as it gives a new degree of freedom to tune the
magnitude of the first updates of a parameter.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [X] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [X] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [X] Add initial_accumulator_value argument to AdaGrad

### CODE REVIEW

**Comment by leezu** [comment]
@eric-haibin-lin

File: python/mxnet/optimizer.py [hunk 1]
```diff
@@ -1091,14 +1091,22 @@ class AdaGrad(Optimizer):
...
+    def __init__(self, eps=1e-7, initial_accumulator_value=0, **kwargs):
         super(AdaGrad, self).__init__(**kwargs)
         self.float_stable_eps = eps
+        self.initial_accumulator_value = initial_accumulator_value
 
     def create_state(self, index, weight):
-        return zeros(weight.shape, weight.context, stype=weight.stype)  # history
+        # history
+        if self.initial_accumulator_value == 0:
+            return zeros(weight.shape, weight.context, stype=weight.stype)
```
Comment (eric-haibin-lin) [inline python/mxnet/optimizer.py]: What about 
```
history = zeros(weight.shape, weight.context, stype=weight.stype)
if self.initial_accumulator_value:
    history[:] = self.initial_accumulator_value
```
This way the stype of history doesn't change

File: python/mxnet/optimizer.py [hunk 2]
```diff
@@ -1091,14 +1091,20 @@ class AdaGrad(Optimizer):
     ----------
     eps: float, optional
         Small value to avoid division by 0.
+    initial_accumulator_value: float, default 0
+        The Adagrad state is initially set to this value.
 
     """
-    def __init__(self, eps=1e-7, **kwargs):
+    def __init__(self, eps=1e-7, initial_accumulator_value=0, **kwargs):
```
Comment (piiswrong) [inline python/mxnet/optimizer.py]: not sure about the name.

Comment (eric-haibin-lin) [reply python/mxnet/optimizer.py]: I think it comes from https://github.com/tensorflow/tensorflow/blob/r1.8/tensorflow/python/training/adagrad.py#L46

**Comment by leezu** [comment]
Tests added

**Comment by eric-haibin-lin** [comment]
I'm not sure if this is necessary. Can the user just specify `eps`?? I agree eps is not a good name but it has the same effect.

**Comment by leezu** [comment]
@eric-haibin-lin can you clarify what you mean is not necessary? I suggest not to use `eps` as initializer for the state as that would change the default behavior (currently we always initialize as 0) and would make it impossible to initialize as `0` as `eps` should never be set to 0.

**Comment by leezu** [comment]
Rebased to rerun tests

**Comment by eric-haibin-lin** [comment]
I mean setting eps to `1+1e-7` seems equivalent to setting initial_accumulator_value to 1
